### ISSUE TITLE
Support Evaluation for ConcatDataset

### ISSUE THREAD

**Issue by Kaeseknacker** [open-mmlab/mmdetection#2907]
It would be nice to evaluate concatenated datasets. This could be used to calculate an "average" performance across different datasets.

Example in config file:

```python
my_data_root  = '/path/to/my/dataset/
my_test_dict = dict(
        type = 'CustomDataset',
        ann_file = [
            my_data_root  + 'ImageSets/subset_1.txt',
            my_data_root  + 'ImageSets/subset_2.txt'
        ],
        img_prefix = my_data_root,
        pipeline = test_pipeline
    )
```

**Where to start:**
`CustomDataset` in [dataset_wrappers.py](https://github.com/open-mmlab/mmdetection/blob/master/mmdet/datasets/dataset_wrappers.py) must implement a evaluate method like this:
```python
def evaluate(self, outputs, args_eval, **kwargs):
    for dataset in self.datasets:
        # TODO: split the outputs
        dataset.evaluate(outputs, args_eval, **kwargs)
```
The problem is, that `output` is larger than the annotation fields stored in the different datasets.

**Comment by ZwwWayne** [comment]
Hi @Kaeseknacker ,
Thanks for your suggestion, you can index the outputs by the length of the dataset. We will take this feature into consideration in the future. PRs are also welcomed if you would like to create one.

**Comment by zimenglan-sysu-512** [comment]
@Kaeseknacker do u solve it?
@ZwwWayne do u have a plan to add this feature?

**Comment by zimenglan-sysu-512** [comment]
@Kaeseknacker 
i try these codes to solve this problem:
```
    def evaluate(self, results, **kwargs):
        # evaluate: assume all datasets use same dataset_type
        assert len(results) == self.cumulative_sizes[-1], \
            f"wrong sizes{self.cumulative_sizes[-1]} v.s. {len(results)}"
        assert "logger" in kwargs
        dataset_index = -1
        eval_results_list = []
        for size, dataset in zip(self.cumulative_sizes, self.datasets):
            start_idx = 0 if dataset_index == -1 else \
                self.cumulative_sizes[dataset_index]
            end_idx = self.cumulative_sizes[dataset_index + 1]
            # need copy.deepcopy() ???
            results_per_dataset = results[start_idx: end_idx]
            kwargs["logger"].info("\nevaluate: {} with {} images now".format(
                dataset.ann_file, len(results_per_dataset)
            ))
            eval_results_per_dataset = dataset.evaluate(
                results_per_dataset, **kwargs
            )
            eval_results_list.append(eval_results_per_dataset)
            dataset_index += 1 #
        # average: dict[str: float|str]
        str_keys = []
        eval_results = eval_results_list[0]
        for eval_results_per_dataset in eval_results_list[1:]:
            for k in eval_results:
                assert k in eval_results_per_dataset
                if type(eval_results[k]) == str: # e.g. bbox_mAP_copypaste
                    if k not in str_keys:
                        str_keys.append(k)
                    continue
                eval_results[k] += eval_results_per_dataset[k]
        # just ignore the str_keys (str: str)
        for k in str_keys:
            eval_results.pop(k)
        kwargs["logger"].info("\nignore some keys: {}".format(
            ", ".join(str_keys).strip()
        ))
        for k in eval_results:
            eval_results[k] /= len(eval_results_list)
        return eval_results
```

**Comment by Kaeseknacker** [comment]
Hey @zimenglan-sysu-512 , thank you very much. It works fine for me! However, I only tested it with mAP.

A few things I noticed:
1.  @zimenglan-sysu-512 : How can I use the logger? kwargs does not contain a logger for me.
2.  args.eval must be given to the method as a parameter (see [here](https://github.com/open-mmlab/mmdetection/blob/3809e5e8b604a841c54a90b01ca9113acb1c27a8/tools/test.py#L145))
3.  @ZwwWayne Unfortunately, I cannot set breakpoints in the evaluate method. These are always skipped. Maybe this is due to MMDataParallel [here](https://github.com/open-mmlab/mmdetection/blob/3809e5e8b604a841c54a90b01ca9113acb1c27a8/tools/test.py#L125)? Does anyone have similar problems? I use PyCharm as an IDE.
4. The mAP value is currently being calculated without including the lengths of the individual data records. Alternatively, one could first combine all the detections and all GTs and then evaluate them once.
